Python/instrumentation.c

```diff
@@ -1281,23 +1300,73 @@ initialize_lines(PyCodeObject *code)
                 else {
                     line_data[i].original_opcode = 0;
                 }
-                if (line >= 0) {
-                    current_line = line;
-                }
+                current_line = line;
         }
         for (int j = 1; j < length; j++) {
             line_data[i+j].original_opcode = 0;
             line_data[i+j].line_delta = NO_LINE;
         }
+        i += length;
+    }
+    for (int i = code->_co_firsttraceable; i < code_len; ) {
+        int opcode = _Py_GetBaseOpcode(code, i);
+        int oparg = 0;
+        while (opcode == EXTENDED_ARG) {
+            oparg = (oparg << 8) | _PyCode_CODE(code)[i].op.arg;
+            i++;
+            opcode = _Py_GetBaseOpcode(code, i);
+        }
+        oparg = (oparg << 8) | _PyCode_CODE(code)[i].op.arg;
+        i += instruction_length(code, i);
```

Comment:
Aside: we do this in enough places that it might make sense to just add a helper function at some point.

Also, I find a do-while a bit easier to follow here (it's all basically the same, though):

```suggestion
        int opcode;
        int oparg = 0;
        do {
            opcode = _Py_GetBaseOpcode(code, i);
            oparg = (oparg << 8) | _PyCode_CODE(code)[i].op.arg;
            i += instruction_length(code, i);
        } while (opcode == EXTENDED_ARG);
```

Reply:
Using a do-while loop would be incorrect, the body of the loop only applies while `opcode == EXTENDED_ARG` which isn't the case for most instructions.